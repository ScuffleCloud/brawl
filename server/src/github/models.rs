use chrono::{DateTime, Utc};
use octocrab::models::pulls::{MergeableState, ReviewState};
use octocrab::models::{InstallationId, IssueState, RepositoryId, RunId, UserId, WorkflowId};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct User {
    pub login: String,
    pub id: UserId,
}

impl Default for User {
    fn default() -> Self {
        Self {
            login: "user".to_string(),
            id: UserId(0),
        }
    }
}

impl From<octocrab::models::UserProfile> for User {
    fn from(value: octocrab::models::UserProfile) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

impl From<octocrab::models::Author> for User {
    fn from(value: octocrab::models::Author) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Label {
    pub name: String,
}

impl From<octocrab::models::Label> for Label {
    fn from(value: octocrab::models::Label) -> Self {
        Self { name: value.name }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct PrBranch {
    pub label: Option<String>,
    pub ref_field: String,
    pub sha: String,
    pub user: Option<User>,
    pub repo: Option<Repository>,
}

impl From<octocrab::models::pulls::Head> for PrBranch {
    fn from(value: octocrab::models::pulls::Head) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

impl From<octocrab::models::pulls::Base> for PrBranch {
    fn from(value: octocrab::models::pulls::Base) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct PullRequest {
    pub number: u64,
    pub title: String,
    pub body: String,
    pub state: Option<IssueState>,
    pub mergeable_state: Option<MergeableState>,
    pub merge_commit_sha: Option<String>,
    pub assignees: Vec<User>,
    pub requested_reviewers: Vec<User>,
    pub user: Option<User>,
    pub labels: Vec<Label>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub draft: Option<bool>,
    pub closed_at: Option<DateTime<Utc>>,
    pub merged_at: Option<DateTime<Utc>>,
    pub head: PrBranch,
    pub base: PrBranch,
}

impl From<octocrab::models::pulls::PullRequest> for PullRequest {
    fn from(value: octocrab::models::pulls::PullRequest) -> Self {
        Self {
            number: value.number,
            title: value.title.unwrap_or_default(),
            body: value.body.unwrap_or_default(),
            state: value.state,
            mergeable_state: value.mergeable_state,
            merge_commit_sha: value.merge_commit_sha,
            assignees: value.assignees.into_iter().flatten().map(|a| a.into()).collect(),
            requested_reviewers: value.requested_reviewers.into_iter().flatten().map(|a| a.into()).collect(),
            user: value.user.map(|u| (*u).into()),
            labels: value.labels.into_iter().flatten().map(|l| l.into()).collect(),
            created_at: value.created_at,
            updated_at: value.updated_at,
            draft: value.draft,
            closed_at: value.closed_at,
            merged_at: value.merged_at,
            head: (*value.head).into(),
            base: (*value.base).into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Review {
    pub user: Option<User>,
    pub state: Option<ReviewState>,
}

impl From<octocrab::models::pulls::Review> for Review {
    fn from(value: octocrab::models::pulls::Review) -> Self {
        Self {
            user: value.user.map(|u| u.into()),
            state: value.state,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Commit {
    pub sha: String,
    pub tree: Option<String>,
}

impl From<octocrab::models::commits::Commit> for Commit {
    fn from(value: octocrab::models::commits::Commit) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.commit.tree.sha),
        }
    }
}

impl From<octocrab::models::commits::GitCommitObject> for Commit {
    fn from(value: octocrab::models::commits::GitCommitObject) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.tree.sha),
        }
    }
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunStatus {
    #[default]
    Queued,
    InProgress,
    Completed,
    Waiting,
    Requested,
    Pending,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunConclusion {
    Success,
    Failure,
    #[default]
    Neutral,
    Cancelled,
    Skipped,
    TimedOut,
    ActionRequired,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
pub struct CheckRunEvent {
    pub id: i64,
    pub name: String,
    pub head_sha: String,
    pub html_url: Option<String>,
    pub details_url: Option<String>,
    pub url: String,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub status: CheckRunStatus,
    pub conclusion: Option<CheckRunConclusion>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Repository {
    pub id: RepositoryId,
    pub name: String,
    pub owner: User,
    pub default_branch: Option<String>,
}

impl From<octocrab::models::Repository> for Repository {
    fn from(value: octocrab::models::Repository) -> Self {
        Self {
            id: value.id,
            name: value.name,
            owner: value.owner.expect("repository has no owner").into(),
            default_branch: value.default_branch,
        }
    }
}

impl Default for Repository {
    fn default() -> Self {
        Self {
            id: RepositoryId(0),
            name: "repo".to_string(),
            owner: User::default(),
            default_branch: Some("main".to_string()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Installation {
    pub id: InstallationId,
    pub account: User,
}

impl Default for Installation {
    fn default() -> Self {
        Self {
            id: InstallationId(0),
            account: User::default(),
        }
    }
}

impl From<octocrab::models::Installation> for Installation {
    fn from(value: octocrab::models::Installation) -> Self {
        Self {
            id: value.id,
            account: value.account.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WorkflowRun {
    pub id: RunId,
    pub workflow_id: WorkflowId,
    pub name: String,
    pub head_sha: String,
    pub head_branch: String,
    pub conclusion: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub status: String,
}

impl From<octocrab::models::workflows::Run> for WorkflowRun {
    fn from(value: octocrab::models::workflows::Run) -> Self {
        Self {
            id: value.id,
            workflow_id: value.workflow_id,
            name: value.name,
            head_sha: value.head_sha,
            head_branch: value.head_branch,
            conclusion: value.conclusion,
            created_at: value.created_at,
            updated_at: value.updated_at,
            status: value.status,
        }
    }
}

impl Default for WorkflowRun {
    fn default() -> Self {
        Self {
            id: RunId(0),
            workflow_id: WorkflowId(0),
            name: "workflow".to_string(),
            head_sha: "sha".to_string(),
            head_branch: "branch".to_string(),
            conclusion: None,
            created_at: Utc::now(),
            updated_at: Utc::now(),
            status: "status".to_string(),
        }
    }
}
