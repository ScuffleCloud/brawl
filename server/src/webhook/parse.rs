use axum::body::Body;
use axum::http::{HeaderMap, HeaderValue, Request, StatusCode};
use hmac::{Hmac, Mac};
use octocrab::models::webhook_events::{EventInstallation, WebhookEventPayload, WebhookEventType};
use octocrab::models::InstallationId;
use sha2::Sha256;

use crate::github::models::{Installation, Repository, User};

fn verify_gh_signature(headers: &HeaderMap<HeaderValue>, body: &[u8], secret: &str) -> bool {
    let Some(signature) = headers.get("x-hub-signature-256").map(|v| v.as_bytes()) else {
        return false;
    };
    let Some(signature) = signature.get(b"sha256=".len()..).and_then(|v| hex::decode(v).ok()) else {
        return false;
    };

    let mut mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).expect("Cannot create HMAC key");
    mac.update(body);
    mac.verify_slice(&signature).is_ok()
}

#[derive(Debug, Clone)]
pub struct WebhookEvent {
    pub sender: Option<User>,
    pub repository: Option<Repository>,
    pub installtion_id: Option<InstallationId>,
    pub installation: Option<Installation>,
    pub kind: WebhookEventType,
    pub specific: WebhookEventPayload,
}

fn parse_event(header: String, body: &[u8]) -> Result<WebhookEvent, (StatusCode, String)> {
    // NOTE: this is inefficient code to simply reuse the code from "derived"
    // serde::Deserialize instead of writing specific deserialization code for the
    // enum.
    let kind: WebhookEventType =
        serde_json::from_value(serde_json::Value::String(header)).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Intermediate structure allows to separate the common fields from
    // the event specific one.
    #[derive(serde::Deserialize)]
    struct Intermediate {
        sender: Option<octocrab::models::Author>,
        repository: Option<octocrab::models::Repository>,
        #[allow(unused)]
        organization: Option<octocrab::models::orgs::Organization>,
        installation: Option<octocrab::models::webhook_events::EventInstallation>,
        #[serde(flatten)]
        specific: serde_json::Value,
    }

    let Intermediate {
        sender,
        repository,
        organization: _,
        installation,
        mut specific,
    } = serde_json::from_slice::<Intermediate>(body).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Bug: OctoCrab wrongly requires the pusher to have an email
    // Remove when https://github.com/XAMPPRocky/octocrab/issues/486 is fixed
    if kind == WebhookEventType::Push {
        if let Some(pusher) = specific.get_mut("pusher") {
            if let Some(email) = pusher.get_mut("email") {
                if email.is_null() {
                    *email = serde_json::Value::String("".to_owned())
                }
            }
        }
    }

    let specific = kind
        .parse_specific_payload(specific)
        .map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    Ok(WebhookEvent {
        sender: sender.map(|s| s.into()),
        repository: repository.map(|r| r.into()),
        installtion_id: match &installation {
            Some(EventInstallation::Full(installation)) => Some(installation.id),
            Some(EventInstallation::Minimal(installation)) => Some(installation.id),
            None => None,
        },
        installation: match installation {
            Some(EventInstallation::Full(installation)) => Some((*installation).into()),
            _ => None,
        },
        kind,
        specific,
    })
}

pub async fn parse_from_request(request: Request<Body>, secret: &str) -> Result<WebhookEvent, (StatusCode, String)> {
    let (parts, body) = request.into_parts();
    let Some(header) = parts.headers.get("X-GitHub-Event").and_then(|v| v.to_str().ok()) else {
        return Err((StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    };

    let Ok(body) = axum::body::to_bytes(body, 1024 * 1024 * 10).await else {
        return Err((StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    };

    if !verify_gh_signature(&parts.headers, &body, secret) {
        return Err((StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    parse_event(header.to_owned(), &body)
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use super::*;

    #[test]
    fn test_parse_event() {
        let event = parse_event("check_run".to_string(), include_bytes!("mock/webhook.check_run.created.json"))
            .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.created", event);

        let event = parse_event(
            "check_run".to_string(),
            include_bytes!("mock/webhook.check_run.completed.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.completed", event);

        let event = parse_event("push".to_string(), include_bytes!("mock/webhook.push.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.push", event);

        let event = parse_event(
            "pull_request".to_string(),
            include_bytes!("mock/webhook.pull_request.synchronize.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request.synchronize", event);

        let event = parse_event(
            "issue_comment".to_string(),
            include_bytes!("mock/webhook.issue_comment.created.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.issue_comment.created", event);

        let event = parse_event("delete".to_string(), include_bytes!("mock/webhook.delete.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.delete", event);

        let event = parse_event("create".to_string(), include_bytes!("mock/webhook.create.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.create", event);

        let event = parse_event(
            "pull_request_review".to_string(),
            include_bytes!("mock/webhook.pull_request_review.submitted.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request_review.submitted", event);
    }

    #[test]
    fn test_verify_gh_signature() {
        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(b"body");
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let mut headers = HeaderMap::new();
        headers.insert("X-Hub-Signature-256", signature.parse().unwrap());
        assert!(verify_gh_signature(&headers, b"body", "secret"));
    }

    #[tokio::test]
    async fn test_parse_from_request() {
        let data = include_bytes!("mock/webhook.push.json");

        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(data);
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let request = Request::builder()
            .header("X-Hub-Signature-256", signature)
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();

        parse_from_request(request, "secret").await.expect("event parsed");
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_signature() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder()
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_missing_header() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder().body(Body::from(data.to_vec())).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_body() {
        let body = Body::from(vec![0; 1024 * 1024 * 20]);
        let request = Request::builder().header("X-GitHub-Event", "push").body(body).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    }
}
