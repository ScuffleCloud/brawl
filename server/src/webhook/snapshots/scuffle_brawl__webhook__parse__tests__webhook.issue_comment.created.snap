---
source: server/src/webhook/parse.rs
expression: event
snapshot_kind: text
---
WebhookEvent {
    sender: Some(
        User {
            login: "scuffle-brawl[bot]",
            id: UserId(
                191062277,
            ),
        },
    ),
    repository: Some(
        Repository {
            id: RepositoryId(
                899726767,
            ),
            name: "ci-testing",
            owner: User {
                login: "ScuffleCloud",
                id: UserId(
                    122814584,
                ),
            },
        },
    ),
    installtion_id: Some(
        InstallationId(
            58286822,
        ),
    ),
    installation: None,
    kind: IssueComment,
    specific: IssueComment(
        IssueCommentWebhookEventPayload {
            action: Created,
            changes: None,
            comment: Comment {
                id: CommentId(
                    2557711955,
                ),
                node_id: "IC_kwDONaC9r86Yc5ZT",
                url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing/issues/comments/2557711955",
                    query: None,
                    fragment: None,
                },
                html_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "github.com",
                        ),
                    ),
                    port: None,
                    path: "/ScuffleCloud/ci-testing/pull/22",
                    query: None,
                    fragment: Some(
                        "issuecomment-2557711955",
                    ),
                },
                issue_url: Some(
                    Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/repos/ScuffleCloud/ci-testing/issues/22",
                        query: None,
                        fragment: None,
                    },
                ),
                body: Some(
                    "ðŸ’” Test failed - [brawl-done](https://github.com/ScuffleCloud/ci-testing/actions/runs/12438404050/job/34730267268)\n",
                ),
                body_text: None,
                body_html: None,
                author_association: Contributor,
                user: Author {
                    login: "scuffle-brawl[bot]",
                    id: UserId(
                        191062277,
                    ),
                    node_id: "BOT_kgDOC2NhBQ",
                    avatar_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "avatars.githubusercontent.com",
                            ),
                        ),
                        port: None,
                        path: "/in/1081680",
                        query: Some(
                            "v=4",
                        ),
                        fragment: None,
                    },
                    gravatar_id: "",
                    url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D",
                        query: None,
                        fragment: None,
                    },
                    html_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "github.com",
                            ),
                        ),
                        port: None,
                        path: "/apps/scuffle-brawl",
                        query: None,
                        fragment: None,
                    },
                    followers_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/followers",
                        query: None,
                        fragment: None,
                    },
                    following_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/following%7B/other_user%7D",
                        query: None,
                        fragment: None,
                    },
                    gists_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/gists%7B/gist_id%7D",
                        query: None,
                        fragment: None,
                    },
                    starred_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/starred%7B/owner%7D%7B/repo%7D",
                        query: None,
                        fragment: None,
                    },
                    subscriptions_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/subscriptions",
                        query: None,
                        fragment: None,
                    },
                    organizations_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/orgs",
                        query: None,
                        fragment: None,
                    },
                    repos_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/repos",
                        query: None,
                        fragment: None,
                    },
                    events_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/events%7B/privacy%7D",
                        query: None,
                        fragment: None,
                    },
                    received_events_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/scuffle-brawl%5Bbot%5D/received_events",
                        query: None,
                        fragment: None,
                    },
                    type: "Bot",
                    site_admin: false,
                    patch_url: None,
                    email: None,
                },
                created_at: 2024-12-20T21:02:02Z,
                updated_at: Some(
                    2024-12-20T21:02:02Z,
                ),
            },
            enterprise: None,
            issue: Issue {
                id: IssueId(
                    2753091006,
                ),
                node_id: "PR_kwDONaC9r86F7taI",
                url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing/issues/22",
                    query: None,
                    fragment: None,
                },
                repository_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing",
                    query: None,
                    fragment: None,
                },
                labels_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing/issues/22/labels%7B/name%7D",
                    query: None,
                    fragment: None,
                },
                comments_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing/issues/22/comments",
                    query: None,
                    fragment: None,
                },
                events_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "api.github.com",
                        ),
                    ),
                    port: None,
                    path: "/repos/ScuffleCloud/ci-testing/issues/22/events",
                    query: None,
                    fragment: None,
                },
                html_url: Url {
                    scheme: "https",
                    cannot_be_a_base: false,
                    username: "",
                    password: None,
                    host: Some(
                        Domain(
                            "github.com",
                        ),
                    ),
                    port: None,
                    path: "/ScuffleCloud/ci-testing/pull/22",
                    query: None,
                    fragment: None,
                },
                number: 22,
                state: Open,
                state_reason: None,
                title: "add fib",
                body: Some(
                    "<!--\r\nThank you for your Pull Request. Please provide a description above and review\r\nthe requirements below.\r\n\r\nBug fixes and new features should include tests.\r\n\r\nContributors guide: https://github.com/ScuffleTV/.github/blob/main/CONTRIBUTING.md\r\n-->\r\n\r\n## Motivation\r\n\r\n<!--\r\nExplain the context and why you're making that change. What is the problem\r\nyou're trying to solve? If a new feature is being added, describe the intended\r\nuse case that feature fulfills.\r\n-->\r\n\r\n## Solution\r\n\r\n<!--\r\nSummarize the solution and provide any necessary context needed to understand\r\nthe code change.\r\n-->\r\n\r\n## Tickets\r\n\r\n<!--\r\nLink any relevant tickets. Remember to not specify the ticket ids in the commit messages.\r\n-->\r\n",
                ),
                body_text: None,
                body_html: None,
                user: Author {
                    login: "TroyKomodo",
                    id: UserId(
                        49777269,
                    ),
                    node_id: "MDQ6VXNlcjQ5Nzc3MjY5",
                    avatar_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "avatars.githubusercontent.com",
                            ),
                        ),
                        port: None,
                        path: "/u/49777269",
                        query: Some(
                            "v=4",
                        ),
                        fragment: None,
                    },
                    gravatar_id: "",
                    url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo",
                        query: None,
                        fragment: None,
                    },
                    html_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "github.com",
                            ),
                        ),
                        port: None,
                        path: "/TroyKomodo",
                        query: None,
                        fragment: None,
                    },
                    followers_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/followers",
                        query: None,
                        fragment: None,
                    },
                    following_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/following%7B/other_user%7D",
                        query: None,
                        fragment: None,
                    },
                    gists_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/gists%7B/gist_id%7D",
                        query: None,
                        fragment: None,
                    },
                    starred_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/starred%7B/owner%7D%7B/repo%7D",
                        query: None,
                        fragment: None,
                    },
                    subscriptions_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/subscriptions",
                        query: None,
                        fragment: None,
                    },
                    organizations_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/orgs",
                        query: None,
                        fragment: None,
                    },
                    repos_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/repos",
                        query: None,
                        fragment: None,
                    },
                    events_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/events%7B/privacy%7D",
                        query: None,
                        fragment: None,
                    },
                    received_events_url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "api.github.com",
                            ),
                        ),
                        port: None,
                        path: "/users/TroyKomodo/received_events",
                        query: None,
                        fragment: None,
                    },
                    type: "User",
                    site_admin: false,
                    patch_url: None,
                    email: None,
                },
                labels: [],
                assignee: None,
                assignees: [],
                author_association: "MEMBER",
                milestone: None,
                locked: false,
                active_lock_reason: None,
                comments: 18,
                pull_request: Some(
                    PullRequestLink {
                        url: Url {
                            scheme: "https",
                            cannot_be_a_base: false,
                            username: "",
                            password: None,
                            host: Some(
                                Domain(
                                    "api.github.com",
                                ),
                            ),
                            port: None,
                            path: "/repos/ScuffleCloud/ci-testing/pulls/22",
                            query: None,
                            fragment: None,
                        },
                        html_url: Url {
                            scheme: "https",
                            cannot_be_a_base: false,
                            username: "",
                            password: None,
                            host: Some(
                                Domain(
                                    "github.com",
                                ),
                            ),
                            port: None,
                            path: "/ScuffleCloud/ci-testing/pull/22",
                            query: None,
                            fragment: None,
                        },
                        diff_url: Url {
                            scheme: "https",
                            cannot_be_a_base: false,
                            username: "",
                            password: None,
                            host: Some(
                                Domain(
                                    "github.com",
                                ),
                            ),
                            port: None,
                            path: "/ScuffleCloud/ci-testing/pull/22.diff",
                            query: None,
                            fragment: None,
                        },
                        patch_url: Url {
                            scheme: "https",
                            cannot_be_a_base: false,
                            username: "",
                            password: None,
                            host: Some(
                                Domain(
                                    "github.com",
                                ),
                            ),
                            port: None,
                            path: "/ScuffleCloud/ci-testing/pull/22.patch",
                            query: None,
                            fragment: None,
                        },
                    },
                ),
                closed_at: None,
                closed_by: None,
                created_at: 2024-12-20T16:44:12Z,
                updated_at: 2024-12-20T21:02:03Z,
            },
        },
    ),
}
